handlers/product-switch-api/src/contributionToSupporterPlus.ts

```diff
@@ -56,56 +65,72 @@ export const switchToSupporterPlus = async (
 	};
 };
 
-const getContributionRefundAmount = (
-	zuoraResponse: ZuoraPreviewResponse,
+export const handleMissingRefundAmount = (
 	catalogInformation: CatalogInformation,
-	termStartDate: Date,
+	subscription: ZuoraSubscription,
+	currentDate: Date,
 ): number => {
-	const termStartDateDayOfMonth = termStartDate.getUTCDate();
-	const dayOfMonthToday = new Date().getUTCDate();
+	const ratePlan = getIfDefined(
+		subscription.ratePlans.find(
+			(ratePlan: RatePlan) =>
+				ratePlan.productRatePlanId ===
+				catalogInformation.contribution.productRatePlanId,
+		),
+		'No matching RatePlan found in Subscription,',
+	);
 
-	const invoice = getIfDefined(
-		zuoraResponse.previewResult?.invoices[0],
-		'No invoice found in the preview response',
+	const chargedThroughDate: Date = getIfDefined(
+		ratePlan.ratePlanCharges.find(
+			(ratePlanCharge: RatePlanCharge) =>
+				ratePlanCharge.productRatePlanChargeId ===
+				catalogInformation.contribution.chargeId,
+		)?.chargedThroughDate,
+		'No matching chargedThroughDate found in Subscription',
 	);
 
-	const contributionRefundAmount = invoice.invoiceItems.find(
-		(invoiceItem) =>
+	if (!(currentDate.toDateString() == chargedThroughDate.toDateString())) {
+		throw Error('No contribution refund amount found in the preview response');
+	}
+
+	return 0;
+};
+
+export const getContributionRefundAmount = (
+	zuoraPreviewInvoice: ZuoraPreviewResponseInvoice,
+	catalogInformation: CatalogInformation,
+	subscription: ZuoraSubscription,
+): number => {
+	const contributionRefundAmount = zuoraPreviewInvoice.invoiceItems.find(
+		(invoiceItem: ZuoraPreviewResponseInvoiceItem) =>
 			invoiceItem.productRatePlanChargeId ===
 			catalogInformation.contribution.chargeId,
 	)?.amountWithoutTax;
 
-	if (!contributionRefundAmount && termStartDateDayOfMonth != dayOfMonthToday) {
-		throw Error('No contribution refund amount found in the preview response');
-	}
-
-	return contributionRefundAmount ? contributionRefundAmount : 0;
+	return (
+		contributionRefundAmount ??
+		handleMissingRefundAmount(catalogInformation, subscription, new Date())
+	);
```

Comment:
not keen on using the word "handle" because it makes it unclear what it's actually doing

I'd be tempted to deal with it as an assertion and then continue as normal if it passes.

The "refundExpected" function is the new name for handleMissingRefundAmount
```suggestion
  if (contributionRefundAmount == undefined && refundExpected(catalogInformation, subscription, new Date())) {
    throw Error('No contribution refund amount found in the preview response');
  }
  
	return (contributionRefundAmount ?? 0);
```